source/src/cip/cipcommon.c

```diff
@@ -1264,6 +1289,146 @@ int DecodePaddedEPath(CipEpath *epath,
+  message_router_response->general_status = kCipErrorSuccess;
+  message_router_response->size_of_additional_status = 0;
+
+  CipClass *class = GetCipClass(message_router_request->request_path.class_id);
+
+  EipStatus internal_state = kEipStatusOk;
+
+  /* Call the PreCreateCallback if the class provides one. */
+  if( NULL != class->PreCreateCallback) {
+    internal_state = class->PreCreateCallback(instance, message_router_request, message_router_response);
+  }
+
+  if (kEipStatusOk == internal_state) {
+    CipInstance *new_instance = AddCipInstances(class, 1); /* add 1 instance to class*/
+    OPENER_ASSERT(NULL != new_instance); /* fail if run out of memory */
+
+    /* Call the PostCreateCallback if the class provides one. */
+    if (NULL != class->PostCreateCallback) {
+      class->PostCreateCallback(new_instance, message_router_request, message_router_response);
+    }
+    OPENER_TRACE_INFO("Instance number %d created\n", new_instance->instance_number);
+  }
+  return kEipStatusOkSend;
+}
+
+EipStatus Delete(CipInstance *RESTRICT const instance,
+                 CipMessageRouterRequest *const message_router_request,
+                 CipMessageRouterResponse *const message_router_response,
+                 const struct sockaddr *originator_address,
+                 const int encapsulation_session) {
+
+  message_router_response->general_status = kCipErrorInstanceNotDeletable;
+  message_router_response->size_of_additional_status = 0;
+  InitializeENIPMessage(&message_router_response->message);
+  message_router_response->reply_service = (0x80 | message_router_request->service);
+
+  EipStatus internal_state = kEipStatusOk;
+
+  CipClass *const class = instance->cip_class;
+
+  /* Call the PreDeleteCallback if the class provides one. */
+  if (NULL != class->PreDeleteCallback) {
+    internal_state = class->PreDeleteCallback(instance, message_router_request,
+                                              message_router_response);
+  }
+
+  if (kEipStatusOk == internal_state) {
+    CipInstance *instances = class->instances;
+
+    // update pointers in instance list
+    instances = class->instances; /* pointer to first instance */
+    if (instances->instance_number ==
+        instance->instance_number) {  // if instance to delete is head
+      class->instances = instances->next;
+    } else {
+      while (NULL !=
+             instances->next)  // as long as what next points to is not zero
+      {
+        CipInstance *next_instance = instances->next;
+        if (next_instance->instance_number == instance->instance_number) {
+          instances->next = next_instance->next;
+          break;
+        }
+        instances = instances->next;
+      }
+    }
+    // free all allocated attributes of instance
+    CipAttributeStruct *attribute =
+        instance->attributes; /* init pointer to array of attributes*/
+    for (int i = 0; i < instance->cip_class->number_of_attributes; i++) {
```

Comment:
the loop counter needs not to be negative. Use the same type as number_of_attributes is